- Problem Statement: 

    - A bike-sharing system is a service in which bikes are made available for shared use to individuals on a
short term basis for a price or free. Many bike share systems allow people to borrow a bike from a "dock"
which is usually computer-controlled wherein the user enters the payment information, and the system
unlocks it. This bike can then be returned to another dock belonging to the same system.

    - A US bike-sharing provider BoomBikes has recently suffered considerable dips in their revenues due to
the ongoing Corona pandemic. The company is finding it very difficut to sustain in the current market
scenario. So, it has decided to come up with a mindful business plan to be able to accelerate its revenue as
soon as the ongoing lockdown comes to an end, and the economy restores to a healthy state.


    - In such an attempt, BoomBikes aspires to understand the demand for shared bikes among the people
after this ongoing quarantine situation ends across the nation due to Covid-19. They have planned this to
prepare themselves to cater to the people's needs once the situation gets better all around and stand out
from other service providers and make huge profits.

    - They have contracted a consulting company to understand the factors on which the demand for these
shared bikes depends. Specially, they want to understand the factors affecting the demand for these
shared bikes in the American market. The company wants to know:

In [105]:
import sqlalchemy 
import pymysql

In [106]:


engine = sqlalchemy.create_engine("mysql+pymysql://root:78275412@Abhisheks-MacBook-Air.local:3306/bike_project")

In [107]:
query = "SELECT * FROM bike_data;"

In [108]:
import pandas as pd 
df = pd.read_sql_query(query, engine)

In [109]:
df.head()

instant      dteday  season  yr  mnth  holiday  weekday  workingday  \
0        1  01-01-2018       1   0     1        0        1           1   
1        2  02-01-2018       1   0     1        0        2           1   
2        3  03-01-2018       1   0     1        0        3           1   
3        4  04-01-2018       1   0     1        0        4           1   
4        5  05-01-2018       1   0     1        0        5           1   

   weathersit       temp     atemp      hum  windspeed  casual  registered  \
0           2  14.110847  18.18125  80.5833  10.749882     331         654   
1           2  14.902598  17.68695  69.6087  16.652113     131         670   
2           1   8.050924   9.47025  43.7273  16.636703     120        1229   
3           1   8.200000  10.60610  59.0435  10.739832     108        1454   
4           1   9.305237  11.46350  43.6957  12.522300      82        1518   

    cnt  
0   985  
1   801  
2  1349  
3  1562  
4  1600

1. Preliminary data analysis 

In [110]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 730 entries, 0 to 729
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     730 non-null    int64  
 1   dteday      730 non-null    object 
 2   season      730 non-null    int64  
 3   yr          730 non-null    int64  
 4   mnth        730 non-null    int64  
 5   holiday     730 non-null    int64  
 6   weekday     730 non-null    int64  
 7   workingday  730 non-null    int64  
 8   weathersit  730 non-null    int64  
 9   temp        730 non-null    float64
 10  atemp       730 non-null    float64
 11  hum         730 non-null    float64
 12  windspeed   730 non-null    float64
 13  casual      730 non-null    int64  
 14  registered  730 non-null    int64  
 15  cnt         730 non-null    int64  
dtypes: float64(4), int64(11), object(1)
memory usage: 91.4+ KB


In [111]:
# 2. Checking for null values 

(
    df 
    .isnull()
    .sum()
)

instant       0
dteday        0
season        0
yr            0
mnth          0
holiday       0
weekday       0
workingday    0
weathersit    0
temp          0
atemp         0
hum           0
windspeed     0
casual        0
registered    0
cnt           0
dtype: int64

- Observations
1. There are 730 rows in the data. 
2. There aren't any missing values in the data. 
3. Data has occupied almost 100kb in the memory. 
4. From the data description and the inital observation of the data it suggests that the `instant` is an index column. So we can drop this as it will not give any valuable information to our data. 
5. `dteday` is a date column which is incorrectly mentioned here as ~object~ type. We will change that into the date time column. 
6. There are four seasons which is mentioned under the season column. 
7. Months of the year is mentioned under the months column
8. `holiday` column explains if the given day is a holiday or a working day. 
9. `weekday` explains the day of the week. This is an ordinal data. 
10. `weathersit` explains the weather of the given day. This is a non ordinal data but it is marked in a way that it is giving sense of an ordinal data. We need to change the values of this category. 
11. `temp` gives the temperature in the given day. 
12. `atemp` is the feels like temperation which could be almost as equal to the `temp`. so there can be a higher collinearity between them. More investigation needed to verify that. 
13. `hum` explains about the humidity in the given day. 
14. `casual` explain about the user. if the user is regular or a casual user. 
15. `registerd` gives that if the user is a registered user or a guest user. 
16. `cnt` total number of rental bike included both casual and registered. 

1.1 instant

In [112]:
# dropping `instant` from the table

(
    df 
    .drop(columns= 'instant')
)

dteday  season  yr  mnth  holiday  weekday  workingday  weathersit  \
0    01-01-2018       1   0     1        0        1           1           2   
1    02-01-2018       1   0     1        0        2           1           2   
2    03-01-2018       1   0     1        0        3           1           1   
3    04-01-2018       1   0     1        0        4           1           1   
4    05-01-2018       1   0     1        0        5           1           1   
..          ...     ...  ..   ...      ...      ...         ...         ...   
725  27-12-2019       1   1    12        0        5           1           2   
726  28-12-2019       1   1    12        0        6           0           2   
727  29-12-2019       1   1    12        0        0           0           2   
728  30-12-2019       1   1    12        0        1           1           1   
729  31-12-2019       1   1    12        0        2           1           2   

          temp     atemp      hum  windspeed  casual  registered   cnt  
0    14.110847  18.18125  80.5833  10.749882     331         654   985  
1    14.902598  17.68695  69.6087  16.652113     131         670   801  
2     8.050924   9.47025  43.7273  16.636703     120        1229  1349  
3     8.200000  10.60610  59.0435  10.739832     108        1454  1562  
4     9.305237  11.46350  43.6957  12.522300      82        1518  1600  
..         ...       ...      ...        ...     ...         ...   ...  
725  10.420847  11.33210  65.2917  23.458911     247        1867  2114  
726  10.386653  12.75230  59.0000  10.416557     644        2451  3095  
727  10.386653  12.12000  75.2917   8.333661     159        1182  1341  
728  10.489153  11.58500  48.3333  23.500518     364        1432  1796  
729   8.849153  11.17435  57.7500  10.374682     439        2290  2729  

[730 rows x 15 columns]

1.2. season 

In [113]:
(
    df 
    .season
    .replace({
        1: "spring", 
        2: "summer", 
        3: "fall", 
        4: "winter"
    })
    
)


0      spring
1      spring
2      spring
3      spring
4      spring
        ...  
725    spring
726    spring
727    spring
728    spring
729    spring
Name: season, Length: 730, dtype: object

1.3. yr 

In [114]:
(
    df 
    .yr
    .value_counts()
)

yr
0    365
1    365
Name: count, dtype: int64

1.4 mnth

In [115]:
(
    df 
    .mnth
    .replace({
        1: "Jan", 
        2: "Feb", 
        3: "Mar", 
        4: "April", 
        5: "May",
        6: "June", 
        7: "July", 
        8: "Aug", 
        9: "Sept", 
        10: "Oct", 
        11: "Nov", 
        12: "Dec" 
        
    })
)

0      Jan
1      Jan
2      Jan
3      Jan
4      Jan
      ... 
725    Dec
726    Dec
727    Dec
728    Dec
729    Dec
Name: mnth, Length: 730, dtype: object

1.5 weekday

In [116]:
(
    df 
    .weekday
    .replace({
        0:"Mon", 
        1: "Tue", 
        2: "Wed", 
        3: "Thu", 
        4: "Fri", 
        5: "Sat", 
        6: "Sun"
    })
)

0      Tue
1      Wed
2      Thu
3      Fri
4      Sat
      ... 
725    Sat
726    Sun
727    Mon
728    Tue
729    Wed
Name: weekday, Length: 730, dtype: object

1.6 weathersit 

In [117]:
(
    df 
    .weathersit
    .replace({
        1: "Clear Skies", 
        2: "Misty Skies", 
        3: "Light Precipitation", 
        4: "Severe Weather"
    })
)

0      Misty Skies
1      Misty Skies
2      Clear Skies
3      Clear Skies
4      Clear Skies
          ...     
725    Misty Skies
726    Misty Skies
727    Misty Skies
728    Clear Skies
729    Misty Skies
Name: weathersit, Length: 730, dtype: object

1.7 date

In [118]:
def treat_date(day):
    return (
        pd 
        .to_datetime(day, dayfirst= True)
    )

Data Cleaning Steps

In [119]:
def clean_data(data):
    return (
        data 
        .drop(columns = ['instant'])
        .assign(
            season = lambda df_: df_['season']
                                .replace({
                                            1: "spring", 
                                            2: "summer", 
                                            3: "fall", 
                                            4: "winter"
                                        }), 
            mnth = lambda df_: df_['mnth']
                                .replace({
                                            1: "Jan", 
                                            2: "Feb", 
                                            3: "Mar", 
                                            4: "April", 
                                            5: "May",
                                            6: "June", 
                                            7: "July", 
                                            8: "Aug", 
                                            9: "Sept", 
                                            10: "Oct", 
                                            11: "Nov", 
                                            12: "Dec" 
                                            
                                        }), 
            weekday = lambda df_: df_['weekday']
                                .replace({
                                            0:"Mon", 
                                            1: "Tue", 
                                            2: "Wed", 
                                            3: "Thu", 
                                            4: "Fri", 
                                            5: "Sat", 
                                            6: "Sun"
                                        }), 
            weathersit = lambda df_: df_['weathersit']
                                .replace({
                                            1: "Clear Skies", 
                                            2: "Misty Skies", 
                                            3: "Light Precipitation", 
                                            4: "Severe Weather"
                                        }), 
            dteday = lambda df_: pd 
                                .to_datetime(df_['dteday'], dayfirst= True)
                                .dt.strftime("%d-%m-%Y"), 
            yr = lambda df_: df_['yr']
                                .replace({
                                    0: '2018', 
                                    1: '2019'
                                }), 
            holiday = lambda df_: df_['holiday']
                                .replace({
                                    0: "No", 
                                    1: "Yes"

                                }),
            workingday = lambda df_: df_['workingday']
                                .replace({
                                    1: "Work Day", 
                                    0: "Non Work Day"
                                })
                                
            
        )

    )

Cleaned Data

In [120]:
bike_cleaned = clean_data(df)

Exporting the cleaned data 

In [121]:
DIR_PATH = "/Users/abhisheksaurav/Desktop/bike_project/data/cleaned_data"
FILE_NAME = "bike_data_cleaned"
bike_cleaned.to_csv(f"{DIR_PATH}/{FILE_NAME}.csv")

Loading the data back into the database 

In [122]:
(
    bike_cleaned 
    .to_sql(
        "bike_cleaned",
        engine, 
        if_exists= "replace" 

    )
)

730

In [123]:
bike_cleaned

dteday  season    yr mnth holiday weekday    workingday   weathersit  \
0    01-01-2018  spring  2018  Jan      No     Tue      Work Day  Misty Skies   
1    02-01-2018  spring  2018  Jan      No     Wed      Work Day  Misty Skies   
2    03-01-2018  spring  2018  Jan      No     Thu      Work Day  Clear Skies   
3    04-01-2018  spring  2018  Jan      No     Fri      Work Day  Clear Skies   
4    05-01-2018  spring  2018  Jan      No     Sat      Work Day  Clear Skies   
..          ...     ...   ...  ...     ...     ...           ...          ...   
725  27-12-2019  spring  2019  Dec      No     Sat      Work Day  Misty Skies   
726  28-12-2019  spring  2019  Dec      No     Sun  Non Work Day  Misty Skies   
727  29-12-2019  spring  2019  Dec      No     Mon  Non Work Day  Misty Skies   
728  30-12-2019  spring  2019  Dec      No     Tue      Work Day  Clear Skies   
729  31-12-2019  spring  2019  Dec      No     Wed      Work Day  Misty Skies   

          temp     atemp      hum  windspeed  casual  registered   cnt  
0    14.110847  18.18125  80.5833  10.749882     331         654   985  
1    14.902598  17.68695  69.6087  16.652113     131         670   801  
2     8.050924   9.47025  43.7273  16.636703     120        1229  1349  
3     8.200000  10.60610  59.0435  10.739832     108        1454  1562  
4     9.305237  11.46350  43.6957  12.522300      82        1518  1600  
..         ...       ...      ...        ...     ...         ...   ...  
725  10.420847  11.33210  65.2917  23.458911     247        1867  2114  
726  10.386653  12.75230  59.0000  10.416557     644        2451  3095  
727  10.386653  12.12000  75.2917   8.333661     159        1182  1341  
728  10.489153  11.58500  48.3333  23.500518     364        1432  1796  
729   8.849153  11.17435  57.7500  10.374682     439        2290  2729  

[730 rows x 15 columns]